Add ability to run tests in a mocha instance multiple times

nicojs opened the pull request:
### Description of the Change

These changes allow mocha to run multiple times with the same test suite programmatically (without reload). It does this with minimal code changes. 

It also adds to validations, for when people call `mocha.run` at the wrong time.

For more details and reasoning, please see https://github.com/mochajs/mocha/issues/2783#issuecomment-616446938

Fixes #2783

<!--

We must be able to understand the design of your change from this description. Keep in mind that the maintainers and/or community members reviewing this PR may not be familiar with the subsystem. Please be verbose.

-->

### Alternate Designs

Alternatively, we could have totally removed all state from tests, suites and runnable, but we don't want to change the entire way of keeping state. Please see #2783 for the entire history. 

<!-- Explain what other alternates were considered and why the proposed version was selected -->

### Why should this be in core?

A lot of people seem interested in running a test suite multiple times without reloading files from disk. See #2783. I personally will use it to [implement mutation switching in Stryker](https://github.com/stryker-mutator/stryker/issues/1514)

### Benefits

Huge performance gain for my mutation testing use case. Other people have valid use cases as well.

### Possible Drawbacks

It should be backward compatible, but since I had to rewrite parts in the core, I might have introduced undesired side effects 🤷‍♂️. Although the integration test suite is awesome!
<!-- What are the possible side-effects or negative impacts of the code change? -->

### Applicable issues

#2783 

### Todo:

- [x] Test: implement `reset`
- [x] Suite: implement `reset`
- [x] Runnable: implement `reset`
- [x] Mocha: implement `dispose` and `cleanReferencesAfterRun`
- [x] Mocha: throw errors when `run` call is invalid.
- [x] Runner: make sure to not leak event listeners.
- [x] Runner: implement `dispose` and `cleanReferencesAfterRun`
- [x] Unit tests
- [x] Integration tests
- [x] Docs

nicojs commented:
>  Coverage increased (+0.2%) to 93.213% 

You're welcome 💁‍♂️

nicojs commented:
@boneskull the documentation website build failed because of an http error, not my doing. Please rerun that job if you can, thanks!

afonsojramos commented:
I will be paying close attention to this PR, for now, it looks good!

boneskull commented:
Huh, I am not sure how to re-deploy this on netlify.  I don't see the option.

boneskull commented on browser-entry.js:
```diff
@@ -52,6 +52,17 @@ process.removeListener = function(e, fn) {
+ * Implements listenerCount for 'uncaughtException'.
+ */
+
+process.listenerCount = function(name) {
```
NOTE:  this is an observation and not a request for changes.

I don't love how `process` is shimmed here.  we might be able to get away with changing the prototype to the `EventEmitter` shim (loaded via browserify via [events](https://npm.im/events), because `process` should be an `EventEmitter`.

nicojs replied:
Yeah I was surprised that that wasn't yet the case.

boneskull commented on lib/errors.js:
```diff
@@ -129,6 +129,35 @@ function createInvalidExceptionError(message, value) {
+/**
+ * Creates an error object to be thrown when a mocha object's `run` method is executed while it is already disposed.
+ * @param {string} message The error message to be displayed.
+ * @param {boolean} cleanReferencesAfterRun the value of `autoDispose`
```
this references `autoDispose` but I don't see this name elsewhere in the file.  should it be `cleanReferencesAfterRun`?

boneskull commented on lib/mocha.js:
```diff
@@ -97,6 +101,7 @@ function Mocha(options) {
   this.options = options;
   // root suite
   this.suite = new exports.Suite('', new exports.Context(), true);
+  this._cleanReferencesAfterRun = true;
```
do you think it'd be better to use a finite state machine here?  when we start dealing with more than a couple boolean flags that get mutated, it starts to get hairy...

nicojs replied:
Yeah, although I think `_cleanReferencesAfterRun` should still be a separate  boolean, since it is independent. The other state bits can be converted I think. Let me see what I can come up with.

nicojs replied:
I've implemented a simple state machine. Will push as soon as local build is green 💚

```js
/**
 * A Mocha instance is a finite state machine.
 * These are the states it can be in.
 */
var mochaStates = utils.defineConstants({
  /**
   * Initial state of the mocha instance
   */
  INIT: 'init',
  /**
   * Mocha instance is running tests
   */
  RUNNING: 'running',
  /**
   * Mocha instance is done running tests and references to test functions and hooks are cleaned.
   * You can reset this state by unloading the test files.
   */
  REFERENCES_CLEANED: 'referencesCleaned',
  /**
   * Mocha instance is disposed and can no longer be used.
   */
  DISPOSED: 'disposed'
});
```

boneskull commented on lib/mocha.js:
```diff
@@ -488,6 +494,39 @@ Mocha.prototype.checkLeaks = function(checkLeaks) {
+ * If disabled, be sure to dispose mocha when you're done to prevent memory leaks.
+ * @public
+ * @see {@link Mocha#dispose}
+ * @default true
```
remove `@default true`

L504 should be something like: 

```
* @param {boolean} [cleanReferencesAfterRun=true] - If `false`, disables dispose
```

boneskull commented on lib/mocha.js:
```diff
@@ -488,6 +494,39 @@ Mocha.prototype.checkLeaks = function(checkLeaks) {
+ * Manually dispose this mocha instance. Mark this instance as `disposed` and unable to run more tests.
+ * It also removes function references to tests functions and hooks, so variables trapped in closures can be cleaned by the garbage collector.
+ * @public
+ * @returns {void}
```
you can remove this line, it's implied

boneskull commented on lib/runner.js:
```diff
@@ -123,16 +123,22 @@ module.exports = Runner;
+ * @param {boolean} [cleanReferencesAfterRun] Whether to clean references to test fns and hooks when a suite is done.
  */
-function Runner(suite, delay) {
+function Runner(suite, delay, cleanReferencesAfterRun) {
```
we're going to be collecting options here (#4245 does it as well).  Let's change the function signature, but in a non-breaking way:

```js
/**
 * Initialize a `Runner` at the Root {@link Suite}, which represents a hierarchy of {@link Suite|Suites} and {@link Test|Tests}.
 *
 * @extends external:EventEmitter
 * @public
 * @class
 * @param {Suite} suite - Root suite
 * @param {Object|boolean} [opts] - Options. If `boolean`, whether or not to delay execution of root suite until ready (for backwards compatibility).
 * @param {boolean} [opts.delay] - Whether to delay execution of root suite until ready.
 * @param {boolean} [opts.cleanReferencesAfterRun] - Whether to clean references to test fns and hooks when a suite is done.
 */
function Runner(suite, opts) {
  var delay;
  if (typeof opts === 'boolean') {
    delay = opts;
    opts = {};
  }
  // etc
```

nicojs replied:
Good point. Will fix.

boneskull commented on lib/runner.js:
```diff
@@ -123,16 +123,22 @@ module.exports = Runner;
   this.failures = 0;
+  /**
+   * @types {[[EventEmitter, string, function]]}
+   */
```
`@types` isn't understood by JSDoc.  afaik there's no way to define a tuple either. the best you can get is `@type {(EventEmitter|string|function)[]}`, I think.

boneskull commented on lib/runner.js:
```diff
@@ -162,6 +168,53 @@ Runner.immediately = global.setImmediate || process.nextTick;
+ * @param fn {function}
+ */
+Runner.prototype._removeEventListener = function(target, eventName, listener) {
+  var eventListenerIndex = this._eventListeners.findIndex(function(
```
This looks to only be used for `process`, but it iterates over everything in `this._eventListeners`.  That doesn't look necessary.  Maybe instead we keep an array of _process_ event listeners (containing only `eventName` and `listener`, or an object with `eventName` key and an array of listener functions as value)--say, `this._processEventListeners`--and instead of naming this function `_removeEventListener`, call it `_removeProcessEventListener`.  In `dispose()`, iterate over this data structure in addition to `this._eventListeners`... assuming removing process listeners in `dispose()` is intended.

nicojs replied:
It is also used for `'error'` listeners on tests and hooks. (L440 and L592)

boneskull replied:
ah, didn't see that.  I am concerned that this is a CPU-bound bottleneck in very large test files / suites.  if it turns out to be, we can refactor it later to be more difficult to understand but more performant. :smile:

boneskull commented on lib/suite.js:
```diff
@@ -92,6 +92,19 @@ function Suite(title, parentContext, isRoot) {
+  });
+  this.tests.forEach(function(test) {
+    test.reset();
+  });
```
what of hooks?

nicojs replied:
I'm not 100% sure I know what you mean with hooks, but I think it is `beforeEach` and friends.

I super duper don't want them to be cleaned. They should stay, just like the test fn's themselves. We're going to need them for the next run.

boneskull replied:
yes, hooks are `beforeEach` etc.  I think I am confused about why suites and tests are getting `reset()` called, but not hooks.

boneskull replied:
to be clear:  I'm still confused.

nicojs replied:
I think I now understand your criticism. Hooks have state as well. And they inherit from `Runnable`. Didn't realize. Will add it when I have some time. Thanks for pointing it out!

boneskull commented on test/unit/runnable.spec.js:
```diff
@@ -179,6 +179,24 @@ describe('Runnable(title, fn)', function() {
+      this._currentRetry = 5;
+      this.pending = true;
+      run.reset();
+      expect(run.timedOut, 'to be', false);
```
fwiw this could be `'to be false'`

boneskull commented on test/integration/fixtures/multiple-runs/multiple-runs-different-output.fixture.js:
```diff
@@ -0,0 +1,31 @@
+const Mocha = require('../../../../lib/mocha');
+
+const mocha = new Mocha({ reporter: 'json' });
+if (process.argv.indexOf('--no-clean-references') >= 0) {
```
Is there precedent for sending these custom arguments elsewhere?  It may have unintended consequences.  a more foolproof could be to modify `env` in the options object (third parameter to `invokeMocha`)

is the reason for these just that you can reuse the fixture in several different ways?  if that's the case, I'd probably prefer just copying the fixtures into separate files, each with their own behavior.  if you need to reuse some code, put that in a common module and `require` it from the fixtures

nicojs replied:
> is the reason for these just that you can reuse the fixture in several different ways

Yes, they are. Will make 3 fixtures for them 👩‍👧‍👧

nicojs commented:
I've implemented (most of) the requested changes. 

I've also added one last bit. When disposed of, a Mocha instancet will now unload files internally. That makes a lot of sense to me. If someone wants to create a new mocha instance in the same process and add the same files again, he should be able to.

boneskull commented:
ty for the update.  looks like there's a conflict.

nicojs commented:
Conflict gone 🤸‍♂️

boneskull commented:
@nicojs there's another conflict 😜

nicojs commented:
You're a busy guy. Fixed once more 🤸‍♂️

nicojs commented:
You're quite good at keeping me busy 🙊

I've implemented the reset on hooks. I saw that `state` and `err` could also be set on a test/hook during a run, so I've added those fields to `Runnable.reset()`. I've also solved the merge conflict.

As far as I know, I've fixed all issues/recommendations.

afonsojramos commented on lib/errors.js:
```diff
@@ -158,6 +158,25 @@ function createMochaInstanceAlreadyRunningError(message, instance) {
+    default:
+      throw new Error('unknown pluginType "' + pluginType + '"');
+  }
+
```
You seem to have missed a closing bracket.

nicojs replied:
Yes, did a merge in the browser 🤷‍♂️

Fixed now. I'm hoping you guys do squash commit? If not, I could do it as well.

boneskull replied:
yes, will squash.

afonsojramos commented:
@boneskull can you take a look if it is good to go? 🚀

boneskull commented:
Seems to be yet another conflict.

boneskull commented:
I would like to call this a "bugfix", since it really always should have worked properly.  But I could see how others may consider this a feature.  Thoughts, @mochajs/core?

boneskull commented:
Something's gone wonky with the changesets here.

I tried to rebase back onto `master`, but that didn't work, and I had more conflicts.  _then_ I did a merge via the GitHub UI to resolve those conflicts, and now I just have...more conflicts.

@nicojs Sorry, I tried to help.  You'll need to pull your branch again from your fork, and pull `master` as well, then retry a rebase back onto `master`.

boneskull commented:
It looks as if there are changesets in here that shouldn't be in here; 
66d9b29 and later.  A rebase will need to ensure those changesets don't end up in this PR.

@nicojs I'll give you a chance, but please let me know if you want me to give it another go.  I may have screwed it up myself!

Might even be best to _not_ pull your branch back down, and instead just retry a rebase onto `master` from your working copy's branch.

nicojs commented:
> Might even be best to not pull your branch back down, and instead just retry a rebase onto master from your working copy's branch.

This is basically what I did.

```shell
git checkout feat/hot-reload
git fetch origin master:master
git rebase master
# Then I fixed the rebase issues that came up and continued rebasing (18 commits). 
git push --force
```

boneskull commented:
thanks, looks like you fixed it.

boneskull commented:
IMO this can land, but want to know if @juergba has any concerns.

afonsojramos commented:
Any updates on this PR @boneskull @juergba ?

Munter commented on lib/runner.js:
```diff
@@ -121,18 +121,30 @@ module.exports = Runner;
+  if (opts === undefined) {
+    opts = {};
+  }
+  if (typeof opts === 'boolean') {
```
It's good that this change is backwards compatible. But if the updated pattern is something we'd like to go ahead with in the future, maybe we should consider a deprecation notice here and queue a removal in a future major release

boneskull replied:
maybe, but we could also just keep it in, since it doesn't look too hairy (at least for now!).  IMO, if/when it becomes a maintenance burden I'd advocate for this approach.

boneskull commented:
@mochajs/core any thoughts on whether this should be a patch or minor?  If our next release will be v8.0.0, I guess it's moot.

nicojs commented:
It should be a minor version IMHO, since it adds a new feature, right?

Munter commented:
Agree on it being a minor release (or rolled into the next major)

boneskull commented:
Will give @juergba 48h to look at this; if he declines to review I'll merge it

boneskull commented:
This is one of those "missing features" that is only a bug because it was missing.  😄

nicojs commented:
True. 

To be fair, I think we same problem in Stryker itself. Creating an instance of `Stryker` and running mutation testing twice is probably not working 🙄 Just don't tell anyone 🤐

nicojs commented:
So... I guess this will be merged soon 🤗

boneskull commented:
yes

boneskull commented:
So it looks like I can't actually use this for the parallel runner.  What I had hoped to do is re-use the same Mocha instance... except with different files (specifically, a single file) on every run.  

@nicojs Wondering if you have an idea about what could be causing this behavior.  What I'm trying to do is:

1. Initialize a `Mocha` instance.  Set `cleanReferencesAfterRun` to `false`.
1. Repeat for _n_ files:
    1. Add a file to `Mocha` instance
    1. call `Mocha#run()`
    1. Upon completion, "reset" Mocha and remove all files from the `Mocha#files` array. 

I attempted to "reset" by calling `dispose()` upon completion (step 2.iii.).  But this fails at the next run, because:

```md
Mocha instance is already disposed, cannot start a new test run. Please create a new mocha instance. Be sure to set disable `cleanReferencesAfterRun` when you want to reuse the same mocha instance for multiple test runs
```

Of course, if I _don't_ call `dispose()`, things get really confused due to extra hooks hanging around.

Why can't I reuse the `Mocha` instance after `dispose()`?

boneskull commented:
Should I just not call `cleanReferencesAfterRun`?  Would it work then?

boneskull commented:
(I suppose I'm rubbber-ducking now...)

boneskull commented:
I think, maybe, that it wasn't the problem you were trying to solve.  but it seems to get most of the way there.

boneskull commented:
OK, so, I am thinking it's really not worth the bother to try to reuse `Mocha` instances in the manner I'm trying (in #4245), even with your code now merged.  Too many things to worry about.  Like dumping files out of the module cache--but we're still running root-level hooks from those files, root-level hooks being applied nondeterministically, etc.  I mean, it's possible, just difficult to do correctly (for some definition of correctly).

anyway, thanks for the PR 😄

nicojs commented:
Yeah, it wasn't the problem I was trying to solve. The reason I added the behavior of throwing an error on run after dispose was just because thought that made the most sense. Feel free to change that behavior, no harm done since it wasn't released yet, right? As long as I can run the exact same suite without reloading files, I'm happy ;)

birtles commented:
This appears to cause breakage for me:
https://github.com/birtles/rikaichamp/runs/707671695#step:5:71
I am getting `createMochaInstanceAlreadyRunningError` when trying to update mocha.

I haven't had a chance to debug why yet. Any tips would be appreciated.

craigtaub commented:
@birtles Please could you submit this bug as a new issue? Please include details of versions as well as how you are calling mocha. Thanks

birtles commented:
> @birtles Please could you submit this bug as a new issue? Please include details of versions as well as how you are calling mocha. Thanks

Hi. I went to submit a bug but the bug form was requiring a little bit more time than I can currently afford, unfortunately. I discovered that dropping a call to `mocha.run()` resolved the issue for me. I have no idea why the call to `mocha.run()` was needed in the first place.

For what it's worth, I'm running with `karma` and `karma-firefox-launcher`.

audioscavenger commented:
i updated the thread https://github.com/mochajs/mocha/issues/2314 and I found out that there must be a bug in the options parser.
by forcing _cleanReferencesAfterRun = false after you instantiated a new mocha, it actually let you re-run your tests! lost 2 hours on this one. You're welcome.
```
var mocha = new Mocha(mochaOptions)
mocha._cleanReferencesAfterRun = false
```
